C#/Botkit/botbuilder-slack-adapter/SlackMessageTypeMiddleware.cs

```diff
@@ -1,22 +1,90 @@
+        public async void OnTurn(TurnContext context, Func<Task<object>> next)
         {
+            if (context.Activity.Type == "message" && context.Activity.ChannelData != null)
+            {
+                var adapter = context.Adapter as SlackAdapter;
 
+                string bot_user_id = await adapter.GetBotUserByTeam(context.Activity);
+                var mentionSyntax = "<@" + bot_user_id + "(\\|.*?)?>";
+                var mention = new Regex(mentionSyntax, RegexOptions.IgnoreCase);
+                var direct_mention = new Regex('^' + mentionSyntax, RegexOptions.IgnoreCase);
+
+                // is this a DM, a mention, or just ambient messages passing through?
+                if ((context.Activity.ChannelData as dynamic)?.channel_type == "im")
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_message";
+
+                    // strip any potential leading @mention
+                    Regex.Replace(
+                        Regex.Replace(
+                            Regex.Replace(
+                                Regex.Replace(context.Activity.Text, direct_mention.ToString(), ""), 
+                                @"/ ^\s +/", ""), 
+                            @"/ ^:\s +/", ""), 
+                        @"/ ^\s +/", "");
+                }
+                else if (!string.IsNullOrEmpty(bot_user_id) && !string.IsNullOrEmpty(context.Activity.Text) && context.Activity.Text.Equals(direct_mention))
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "direct_mention";
+
+                    // strip the @mention
+                    Regex.Replace(
+                        Regex.Replace(
+                            Regex.Replace(
+                                Regex.Replace(context.Activity.Text, direct_mention.ToString(), ""),
+                                @"/ ^\s +/", ""),
+                            @"/ ^:\s +/", ""),
+                        @"/ ^\s +/", "");
+                }
+                else if (!string.IsNullOrEmpty(bot_user_id) && string.IsNullOrEmpty(context.Activity.Text) && context.Activity.Text.Equals(mention))
+                {
+                    (context.Activity.ChannelData as dynamic).botkitEventType = "mention";
+                }
+                // else, this is an "ambient" message
+
+                // if this is a message from a bot, we probably want to ignore it.
+                // switch the botkit event type to bot_message
+                // and the activity type to Event <-- will stop it from being included in dialogs
+                // NOTE: This catches any message from any bot, including this bot.
+                // Note also, bot_id here is not the same as bot_user_id so we can't (yet) identify messages originating from this bot without doing an additional API call.
+                if ((context.Activity.ChannelData as dynamic)?.bot_id != null)
```

Comment:
Replace `bot_id` with `botId`, but we must be careful to also change where this is assigned